common/modules/original_owner_helper/src/lib.rs

```diff
@@ -21,7 +28,7 @@ pub trait OriginalOwnerHelperModule {
             } else {
                 require!(
                     original_owner == payment_original_owner,
-                    "All position must have the same original owner"
+                    "Original owner is not the same for all payments"
```

Comment:
Can't comment on the whole code, but instead of using `ManagedAddress::zero()` as the default value, you could do this to make it more clear:

```
let mut opt_original_owner = None;

//// other code

for ...
{
    /////

    match opt_original_owner {
        Some(original_owner) => require!(...),
        None => opt_original_owner = Some(payment_original_owner),
    }
}
```